File: lib/java/src/main/java/io/github/brenbar/uapi/internal/GenerateRandomUnion.java
package io.github.brenbar.uapi.internal;

import java.util.HashMap;
import java.util.List;
import java.util.Map;

import io.github.brenbar.uapi.RandomGenerator;
import io.github.brenbar.uapi.internal.types.UStruct;
import io.github.brenbar.uapi.internal.types.UTypeDeclaration;

import static io.github.brenbar.uapi.internal.ConstructRandomUnion.constructRandomUnion;

public class GenerateRandomUnion {
    public static Object generateRandomUnion(Object blueprintValue, boolean useBlueprintValue,
            boolean includeOptionalFields, boolean randomizeOptionalFields, List<UTypeDeclaration> typeParameters,
            List<UTypeDeclaration> generics,
            RandomGenerator randomGenerator, Map<String, UStruct> cases) {
        if (useBlueprintValue) {
            final var startingUnionCase = (Map<String, Object>) blueprintValue;
            return constructRandomUnion(cases, startingUnionCase, includeOptionalFields, randomizeOptionalFields,
                    typeParameters, randomGenerator);
        } else {
            return constructRandomUnion(cases, new HashMap<>(), includeOptionalFields, randomizeOptionalFields,
                    typeParameters, randomGenerator);
        }
    }
}

File: lib/ts/uapi/internal/generation/GenerateRandomUnion.ts
import { RandomGenerator } from 'uapi/RandomGenerator';
import { UStruct } from 'uapi/internal/types/UStruct';
import { UTypeDeclaration } from 'uapi/internal/types/UTypeDeclaration';
import { constructRandomUnion } from 'uapi/internal/generation/ConstructRandomUnion';

export function generateRandomUnion(
    blueprintValue: any,
    useBlueprintValue: boolean,
    includeOptionalFields: boolean,
    randomizeOptionalFields: boolean,
    typeParameters: UTypeDeclaration[],
    generics: UTypeDeclaration[],
    randomGenerator: RandomGenerator,
    cases: { [key: string]: UStruct },
): any {
    if (useBlueprintValue) {
        const startingUnionCase = blueprintValue as { [key: string]: any };
        return constructRandomUnion(
            cases,
            startingUnionCase,
            includeOptionalFields,
            randomizeOptionalFields,
            typeParameters,
            randomGenerator,
        );
    } else {
        return constructRandomUnion(
            cases,
            {},
            includeOptionalFields,
            randomizeOptionalFields,
            typeParameters,
            randomGenerator,
        );
    }
}
